### ISSUE TITLE
MockResourceArgs.Inputs is empty in NewResourceAsync() for ComponentResource

### ISSUE THREAD

**Issue by ie-zero** [pulumi/pulumi#7822]
In .NET, I attempted to create some parameterized for our custom resources. It appears that the `MockResourceArgs.Inputs` is always an empty collection for all the classes extending `ComponentResource`. The same code works fine for the native Pulumi resources.

I believe the custom resources and the argument classes have been defined correctly. Am I missing something or it is not currently possible to do this?

## Software Versions
Pulumi CLI = v3.10.1
.NET Core = 3.1

**NuGet Packages**
Pulumi = 3.*
Pulumi.AzureNative = 1.22.0

## Example Code
``` csharp
namespace Example
{
  public sealed class MyResourceGroup : ComponentResource
  {
    [Output("name")] public Output<string>? Name { get; private set; }

    public MyResourceGroup(string name, MyResourceGroupArgs args, ComponentResourceOptions? options = null)
      : base("example:ResourceGroup", name, options)
    {
      var resourceGroup = new ResourceGroup($"ResourceGroup-{name}", new ResourceGroupArgs
      {
        Location = args.Location,
        ResourceGroupName = Output.Format($"PULUMI-{args.NameSuffix}")
      });

      Name = resourceGroup.Name;
    }
  }
}

namespace PulumiBug
{
  public sealed class MyResourceGroupMocks : IMocks
  {
    public Task<object> CallAsync(MockCallArgs args)
    {
      var outputs = ImmutableDictionary.CreateBuilder<string, object>();
      return Task.FromResult((object)outputs);
    }

    public Task<(string? id, object state)> NewResourceAsync(MockResourceArgs args)
    {
      var outputs = ImmutableDictionary.CreateBuilder<string, object>();

      // Forward all input parameters as resource outputs, so that we could test them.
      outputs.AddRange(args.Inputs);

      if (args.Type == "example:ResourceGroup")
      {
        outputs.Add("nameSuffix", "XYZ");
      }

      // Default the resource ID to `{name}_id`.
      // We could also format it as `/subscription/abc/resourceGroups/xyz/...` if that was important for tests.
      args.Id ??= $"{args.Name}_id";
      return Task.FromResult((args.Id, (object)outputs))!;
    }

    public static Task<ImmutableArray<Resource>> RunAsync<T>() where T : Stack, new()
    {
      return Deployment.TestAsync<T>(new MyResourceGroupMocks(), new TestOptions { IsPreview = false });
    }
  }

  public sealed class ComplexTests
  {
    private sealed class ResourceGroupStack : Stack
    {
      public ResourceGroupStack()
      {
        _ = new MyResourceGroup("RG", new MyResourceGroupArgs
        {
          Location = "NorthEurope",
          NameSuffix = "ANY_NAME",
        });

        _ = new ResourceGroup("ResourceGroup-Static", new ResourceGroupArgs
        {
          Location = "WestEurope",
          ResourceGroupName = "PULUMI-STATIC"
        });
      }
    }

    [Fact]
    public async Task AnyTest()
    {
      var resources = await MyResourceGroupMocks.RunAsync<ResourceGroupStack>();

      var rg = resources.OfType<ResourceGroup>().First();

      var name = await rg.Name.GetValueAsync();
      Assert.Equal("AllowAzureMonitorOutbound", name);
    }
  }
}
```

**Expected:** The `args` to contain the values passed `NewResourceAsync` i.e. `Location` and `NameSuffix` for `MyResourceGroup`
**Actual:** The `args` is an empty collection for `MyResourceGroup`

**Comment by ie-zero** [comment]
Our main aim is parameterize the unit tests so we avoid creating multiple Stacks (one for each test case). If there is a better way to achieve the same result it would be great.

**Comment by pgavlin** [comment]
The issue at hand here is that `MyResourceGroup` is a component resource rather than a custom resource. The inputs for a component resource aren't used by the engine, so they're not present in mocks (indeed, the outputs for a component resource are also ignored).

> Our main aim is parameterize the unit tests so we avoid creating multiple Stacks (one for each test case). If there is a better way to achieve the same result it would be great.

I'd suggest parameterizing based on the resource name.

**Comment by wichon** [comment]
I am having a similar issue in Python with my unit tests, I need to build a Docker image inside a ComponentResource and pushit it to ECR, so I am mocking the `get_credentials` call from the ECR package like:

```
    def call(self, args: pulumi.runtime.MockCallArgs):
        if (args.token == 'aws:ecr/getCredentials:getCredentials'):
            return {
                'authorization_token' : base64.b64encode('user:pass'.encode('ascii')).decode('ascii'),
                'proxy_endpoint': 'example.com'
            }
        return {}
```

The call is actually being mocked, but the values are not being passed to the `AwaitableGetCredentialsResult` object, all attributes are set to none.

This is how the Docker image is being published to an ECR authenticated repo in the ComponentResource class:
```
  def __get_docker_registry_credentials(self, registryId):
        creds = ecr.get_credentials(registry_id = registryId)
        decoded = base64.b64decode(creds.authorization_token).decode()
        parts = decoded.split(':')
        if len(parts) != 2:
            raise Exception("Invalid credentials")
        return docker.ImageRegistry(creds.proxy_endpoint, parts[0], parts[1])

    def provision_docker_image(self, ecrRepository):
        dockerRegistry = ecrRepository.registry_id.apply(self.__get_docker_registry_credentials)

        dockerImage = docker.Image(
            name = 'web-app-docker-image',
            image_name = ecrRepository.repository_url,
            build = self.stackContext.dockerPath,
            skip_push = False,
            registry = dockerRegistry,
            opts = ResourceOptions(parent = self)
        )
        self.outputs['ecrImageUrl'] = ecrRepository.repository_url
        return dockerImage
```

**Comment by ie-zero** [comment]
> The issue at hand here is that `MyResourceGroup` is a component resource rather than a custom resource. The inputs for a component resource aren't used by the engine, so they're not present in mocks (indeed, the outputs for a component resource are also ignored).
> 
> > Our main aim is parameterize the unit tests so we avoid creating multiple Stacks (one for each test case). If there is a better way to achieve the same result it would be great.
> 
> I'd suggest parameterizing based on the resource name.

Hi @pgavlin, 

Thank you for your response but the aim is to test our logic not the Pulumi resources. Also there are cases which we compose `ComponentResource` from other `ComponentResources` which in that case capturing the input/outputs will be certainly useful.

If there are no plans to support this scenario, is there any example on how to create our own custom resource in C#?

**Comment by pgavlin** [comment]
> ```
>     def call(self, args: pulumi.runtime.MockCallArgs):
>         if (args.token == 'aws:ecr/getCredentials:getCredentials'):
>             return {
>                 'authorization_token' : base64.b64encode('user:pass'.encode('ascii')).decode('ascii'),
>                 'proxy_endpoint': 'example.com'
>             }
>         return {}
> ```
> ...
> The call is actually being mocked, but the values are not being passed to the AwaitableGetCredentialsResult object, all attributes are set to none.

@wichon I think that the problem is that the properties returned by the mock are in `snake_case`, but the engine + SDKs expect properties in `camelCase`. I _think_ that the following will work:

> ```
>     def call(self, args: pulumi.runtime.MockCallArgs):
>         if (args.token == 'aws:ecr/getCredentials:getCredentials'):
>             return {
>                 'authorizationToken' : base64.b64encode('user:pass'.encode('ascii')).decode('ascii'),
>                 'proxyEndpoint': 'example.com'
>             }
>         return {}
> ```

@emiliza we should make sure to add a note about this to our docs.

**Comment by pgavlin** [comment]
> Hi @pgavlin,
> 
> Thank you for your response but the aim is to test our logic not the Pulumi resources. Also there are cases which we compose ComponentResource from other ComponentResources which in that case capturing the input/outputs will be certainly useful.
> 
> If there are no plans to support this scenario, is there any example on how to create our own custom resource in C#?

Perhaps I'm misunderstanding your scenario. Can you explain exactly what you're trying to do? I was responding based on the scenario you described here:

> Our main aim is parameterize the unit tests so we avoid creating multiple Stacks (one for each test case). If there is a better way to achieve the same result it would be great.

**Comment by wichon** [comment]
> > ```
> >     def call(self, args: pulumi.runtime.MockCallArgs):
> >         if (args.token == 'aws:ecr/getCredentials:getCredentials'):
> >             return {
> >                 'authorization_token' : base64.b64encode('user:pass'.encode('ascii')).decode('ascii'),
> >                 'proxy_endpoint': 'example.com'
> >             }
> >         return {}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ...
> > The call is actually being mocked, but the values are not being passed to the AwaitableGetCredentialsResult object, all attributes are set to none.
> 
> @wichon I think that the problem is that the properties returned by the mock are in `snake_case`, but the engine + SDKs expect properties in `camelCase`. I _think_ that the following will work:
> 
> > ```
> >     def call(self, args: pulumi.runtime.MockCallArgs):
> >         if (args.token == 'aws:ecr/getCredentials:getCredentials'):
> >             return {
> >                 'authorizationToken' : base64.b64encode('user:pass'.encode('ascii')).decode('ascii'),
> >                 'proxyEndpoint': 'example.com'
> >             }
> >         return {}
> > ```
> 
> @emiliza we should make sure to add a note about this to our docs.

That worked, thanks!

**Comment by joeduffy** [comment]
I too am surprised by this behavior. Indeed, I'm surprised we took the path we did, for this reason (among others like wanting to see component inputs and properties in the service console, being available to run Policy as Code rules against, etc). Funny enough, multiple years later I guess I haven't changed my opinion on this, as I had described this same concern way back when: https://github.com/pulumi/pulumi/pull/2296#issuecomment-447709691.

The culprit here is: https://github.com/pulumi/pulumi/blob/6723272dcf4cbd25f4e26ba9cbdd2fca4daa9d30/sdk/nodejs/resource.ts#L860-L868
